### ISSUE TITLE
worker dies after a (long) inactivity

### ISSUE THREAD

**Issue by ELE3E** [rq/rq#915]
Hi I have a setup that has a number of workers listening for on demand jobs. After a while of inactivity, i.e. a list of messages like the one below, the heartbeat signals stop appearing and the worker is not available any more in list of workers (I have an RQ dashboard setup as well).
The time after it stops is not consistent, nor is the number of heartbeats sent before dying.
The process of the worker that died is still there.
How can I check what the root cause of the problem is?

> Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.

**Comment by selwin** [comment]
Is the rqworker process still running? Is there any failed job in the queue that causes the worker to crash?

You should also use a process manager like systemd or supervisor to run RQ in production.

**Comment by ELE3E** [comment]
Yes the rq worker process is still running. It was not executing a job at that time, a long list of heartbeat signals before it stopped.
We are running it on a docker system, as soon as the process exits it will restart automatically.

**Comment by selwin** [comment]
Do you have a job with infinite timeout?

**Comment by ELE3E** [comment]
No I do not, always take the default timeout, but the jobs are usually done in a few seconds

**Comment by samuelcolvin** [comment]
Are there any changes in the memory usage of that container?

Can you see network traffic for that container? If so what is happening

**Comment by ELE3E** [comment]
No network traffic, constant memory usage.

On 10 Jan 2018, at 16:08, Samuel Colvin <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #915<https://github.com/rq/rq/issues/915>.

**Comment by samuelcolvin** [comment]
You might want to enable `debug` logging to see if it shows any more details (if you haven't already).

Also, could this be the worker stopping [here](https://github.com/rq/rq/blob/afbbe47dbbf776a7729b4e2704363484e0eb6c93/rq/worker.py#L492) perhaps due to hitting `default_worker_ttl` or a problem getting a job from the queue?

**Comment by ELE3E** [comment]
Thanks for the tips. I will put the logging level to debug and also enable the verbose option. Unfortunely, I will have to wait until the worker halts again.
From my logs, I don't see any job around the time the worker halted, so I don't think the problem is getting a job from the queue. 
What is this default_worker_ttl for?

**Comment by vithickvi** [comment]
I am still facing this issue. Have 5 worker servers, if there is no job for a couple of hours then automatically the worker server disappears from rq info. Every time i have to go and restart the supervisor in each worker servers

**Comment by nzjrs** [comment]
I have workers that run for weeks with no error, some waiting days for a job.

**Comment by samuelcolvin** [comment]
@vithickvi are the redis keys associated with the worker expiring? Can you check if the [ttl](https://redis.io/commands/ttl) of those keys in redis?

**Comment by ELE3E** [comment]
I’ve noticed that sometimes a worker remains active (and accepting jobs), but are not visible anymore in the info.

On 13 Mar 2018, at 12:34, Samuel Colvin <notifications@github.com<mailto:notifications@github.com>> wrote:


@vithickvi<https://github.com/vithickvi> are the redis keys associated with the worker expiring? Can you check if the ttl<https://redis.io/commands/ttl> of those keys in redis?

**Comment by dralley** [comment]
Not sure if issues are closed when the release goes out or when the patch gets merged, but this can be closed after #945
